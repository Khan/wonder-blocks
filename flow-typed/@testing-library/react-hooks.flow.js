// @flow
/**
 * This file is based on:
 * - https://github.com/testing-library/react-hooks-testing-library/blob/main/src/types/index.ts
 * - https://github.com/testing-library/react-hooks-testing-library/blob/main/src/types/react.ts
 * - https://github.com/testing-library/react-hooks-testing-library/blob/main/src/pure.ts
 */
 declare module "@testing-library/react-hooks" {
    declare export type Renderer<TProps> = {|
        render: (props?: TProps) => void,
        rerender: (props?: TProps) => void,
        unmount: () => void,
        act: Act,
    |};

    declare export type ServerRenderer<TProps> = {|
        ...Renderer<TProps>,
        hydrate: () => void,
    |};

    declare export type RendererProps<TProps, TResult> = {|
        callback: (props: TProps) => TResult,
        setError: (error: Error) => void,
        setValue: (value: TResult) => void,
    |};

    declare export type CreateRenderer<
        TProps,
        TResult,
        TRendererOptions: {...},
        TRenderer: Renderer<TProps>,
    > = (
        props: RendererProps<TProps, TResult>,
        options: TRendererOptions,
    ) => TRenderer;

    declare export type RenderResult<TValue> = {|
        +all: Array<TValue | Error>,
        +current: TValue,
        +error?: Error,
    |};

    declare export type ResultContainer<TValue> = {|
        result: RenderResult<TValue>,
    |};

    declare export type WaitOptions = {|
        interval?: number | false,
        timeout?: number | false,
    |};

    declare export type WaitForOptions = WaitOptions;
    declare export type WaitForValueToChangeOptions = WaitOptions;
    declare export type WaitForNextUpdateOptions = WaitOptions["timeout"];

    declare export type WaitFor = (
        callback: () => boolean | void,
        options?: WaitForOptions,
    ) => Promise<void>;
    declare export type WaitForValueToChange = (
        selector: () => mixed,
        options?: WaitForValueToChangeOptions,
    ) => Promise<void>;
    declare export type WaitForNextUpdate = (
        options?: WaitForNextUpdateOptions,
    ) => Promise<void>;

    declare export type AsyncUtils = {|
        waitFor: WaitFor,
        waitForValueToChange: WaitForValueToChange,
        waitForNextUpdate: WaitForNextUpdate,
    |};

    declare export type RenderHookResult<
        TProps,
        TValue,
        TRenderer: Renderer<TProps> = Renderer<TProps>,
    > = {|
        ...ResultContainer<TValue>,
        // Omit<Renderer<TProps>, 'render' | 'act'>
        ...$Diff<
            Renderer<TProps>,
            {|
                render: Renderer<TProps>["render"],
                act: Renderer<TProps>["act"],
            |},
        >,
        // Omit<TRenderer, keyof Renderer<TProps>>
        ...$Diff<TRenderer, Renderer<TProps>>,
        ...AsyncUtils,
    |};

    declare export type ServerRenderHookResult<
        TProps,
        TValue,
        TRenderer: ServerRenderer<TProps> = ServerRenderer<TProps>,
    > = RenderHookResult<TProps, TValue, TRenderer>;

    declare export type Act = {|
        (callback: () => Promise<void>): Promise<void>,
        (callback: () => void): void,
    |};

    declare export type CleanupCallback = () => Promise<void> | void;

    declare export type WrapperComponent<TProps> = React$ComponentType<TProps>;

    declare export type RendererOptions<TProps> = {|
        wrapper?: WrapperComponent<TProps>,
    |};

    declare export type RenderHookOptions<TProps> = {|
        initialProps?: TProps,
        wrapper?: WrapperComponent<any>,
    |};

    declare export type ReactHooksRenderer = {|
        renderHook: <TProps, TResult>(
            callback: (props: TProps) => TResult,
            options?: RenderHookOptions<TProps>,
        ) => RenderHookResult<TProps, TResult>,
        act: Act,
        cleanup: () => Promise<void>,
        addCleanup: (callback: CleanupCallback) => () => void,
        removeCleanup: (callback: CleanupCallback) => void,
        suppressErrorOutput: () => () => void,
    |};

    declare export type ReactHooksServerRenderer = {|
        // Omit<ReactHooksRenderer, 'renderHook'>
        ...$Diff<
            ReactHooksRenderer,
            {|
                renderHook: ReactHooksRenderer["renderHook"],
            |},
        >,
        renderHook: <TProps, TResult>(
            callback: (props: TProps) => TResult,
            options?: RenderHookOptions<TProps>,
        ) => ServerRenderHookResult<TProps, TResult>,
    |};

    declare export function renderHook<TProps, TResult>(
        callback: (props: TProps) => TResult,
        options?: RenderHookOptions<TProps>,
    ): RenderHookResult<TProps, TResult>;
    declare export var act: Act;
    declare export function cleanup(): Promise<void>;
    declare export function addCleanup(callback: CleanupCallback): () => void;
    declare export function removeCleanup(callback: CleanupCallback): void;
    declare export function suppressErrorOutput(): () => void;
}
