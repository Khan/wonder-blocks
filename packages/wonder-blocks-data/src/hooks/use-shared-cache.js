// @flow
import * as React from "react";
import {DataError, DataErrors} from "../util/data-error.js";
import {ScopedInMemoryCache} from "../util/scoped-in-memory-cache.js";
import type {ValidCacheData, ScopedCache} from "../util/types.js";

/**
 * A function for inserting a value into the cache or clearing it.
 */
type CacheValueFn<TValue: ValidCacheData> = (value: ?TValue) => void;

/**
 * This is the cache.
 * It's incredibly complex.
 * Very in-memory. So cache. Such complex. Wow.
 */
const cache = new ScopedInMemoryCache();

/**
 * Access to the shared in-memory cache.
 *
 * This is the cache used by `useSharedCache` and related hooks and
 * components.
 */
export const SharedCache: ScopedCache = cache;

/**
 * Hook to retrieve data from and store data in an in-memory cache.
 *
 * @returns {[?ReadOnlyCacheValue, CacheValueFn]}
 * Returns an array containing the current cache entry (or undefined), a
 * function to set the cache entry (passing null or undefined to this function
 * will delete the entry).
 *
 * NOTE: Unlike useState or useReducer, we don't automatically update folks
 * if the value they reference changes. We might add it later (if we need to),
 * but the likelihood here is that things won't be changing in this cache in a
 * way where we would need that. If we do (and likely only in specific
 * circumstances), we should consider adding a simple boolean useState that can
 * be toggled to cause a rerender whenever the referenced cached data changes
 * so that callers can re-render on cache changes. However, we should make
 * sure this toggling is optional - or we could use a callback argument, to
 * achieve this on an as-needed basis.
 */
export const useSharedCache = <TValue: ValidCacheData>(
    id: string,
    scope: string,
    initialValue?: ?TValue | (() => ?TValue),
): [?TValue, CacheValueFn<TValue>] => {
    // Verify arguments.
    if (!id || typeof id !== "string") {
        throw new DataError(
            "id must be a non-empty string",
            DataErrors.InvalidInput,
        );
    }

    if (!scope || typeof scope !== "string") {
        throw new DataError(
            "scope must be a non-empty string",
            DataErrors.InvalidInput,
        );
    }

    // Memoize our APIs.
    // This one allows callers to set or replace the cached value.
    const cacheValue = React.useCallback(
        (value: ?TValue) =>
            value == null
                ? cache.purge(scope, id)
                : cache.set(scope, id, value),
        [id, scope],
    );

    // We don't memo-ize the current value, just in case the cache was updated
    // since our last run through. Also, our cache does not know what type it
    // stores, so we have to cast it to the type we're exporting. This is a
    // dev time courtesy, rather than a runtime thing.
    // $FlowIgnore[incompatible-type]
    let currentValue: ?TValue = cache.get(scope, id);

    // If we have an initial value, we need to add it to the cache
    // and use it as our current value.
    if (currentValue == null && initialValue !== undefined) {
        // Get the initial value.
        const value =
            typeof initialValue === "function" ? initialValue() : initialValue;

        if (value != null) {
            // Update the cache.
            cacheValue(value);

            // Make sure we return this value as our current value.
            currentValue = value;
        }
    }

    // Now we have everything, let's return it.
    return [currentValue, cacheValue];
};
