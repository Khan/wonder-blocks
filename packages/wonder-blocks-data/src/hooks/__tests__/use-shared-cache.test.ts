// eslint-disable-next-line import/no-unassigned-import
import "jest-extended";
import {renderHook} from "@testing-library/react";

import {useSharedCache, SharedCache} from "../use-shared-cache";

describe("#useSharedCache", () => {
    beforeEach(() => {
        SharedCache.purgeAll();
    });

    it.each`
        id
        ${null}
        ${""}
        ${5}
        ${() => "BOO"}
    `("should throw if the id is $id", ({id}: any) => {
        // Arrange

        // Act
        const underTest = () => renderHook(() => useSharedCache(id, "scope"));

        // Assert
        expect(underTest).toThrowErrorMatchingSnapshot();
    });

    it.each`
        scope
        ${null}
        ${""}
        ${5}
        ${() => "BOO"}
    `("should throw if the scope is $scope", ({scope}: any) => {
        // Arrange

        // Act
        const underTest = () => renderHook(() => useSharedCache("id", scope));

        // Assert
        expect(underTest).toThrowErrorMatchingSnapshot();
    });

    it("should return a tuple of two items", () => {
        // Arrange

        // Act
        const {
            result: {current: result},
        } = renderHook(() => useSharedCache("id", "scope"));

        // Assert
        expect(result).toBeArrayOfSize(2);
    });

    describe("tuple[0] - currentValue", () => {
        it("should be null if nothing is cached", () => {
            // Arrange

            // Act
            const {
                result: {current: result},
            } = renderHook(() => useSharedCache("id", "scope"));

            // Assert
            expect(result[0]).toBeNull();
        });

        it("should match initialValue when provided as a non-function", () => {
            // Arrange

            // Act
            const {
                result: {current: result},
            } = renderHook(() =>
                useSharedCache("id", "scope", "INITIAL VALUE"),
            );

            // Assert
            expect(result[0]).toBe("INITIAL VALUE");
        });

        it("should match the return of initialValue when provided as non-function", () => {
            // Arrange

            // Act
            const {
                result: {current: result},
            } = renderHook(() =>
                useSharedCache("id", "scope", () => "INITIAL VALUE"),
            );

            // Assert
            expect(result[0]).toBe("INITIAL VALUE");
        });
    });

    describe("tuple[1] - setValue", () => {
        it("should be a function", () => {
            // Arrange

            // Act
            const {
                result: {current: result},
            } = renderHook(() => useSharedCache("id", "scope"));

            // Assert
            expect(result[1]).toBeFunction();
        });

        it("should be the same function if the id and scope remain the same", () => {
            // Arrange
            const wrapper = renderHook(
                ({id, scope}: any) => useSharedCache(id, scope),
                {initialProps: {id: "id", scope: "scope"}},
            );

            // Act
            wrapper.rerender({
                id: "id",
                scope: "scope",
            });
            const value1 = wrapper.result.all[wrapper.result.all.length - 2];
            const value2 = wrapper.result.current;
            const result1 = Array.isArray(value1) ? value1[1] : "BAD1";
            const result2 = Array.isArray(value2) ? value2[1] : "BAD2";

            // Assert
            expect(result1).toBe(result2);
        });

        it("should be a new function if the id changes", () => {
            // Arrange
            const wrapper = renderHook(
                ({id}: any) => useSharedCache(id, "scope"),
                {
                    initialProps: {id: "id"},
                },
            );

            // Act
            wrapper.rerender({id: "new-id"});
            const value1 = wrapper.result.all[wrapper.result.all.length - 2];
            const value2 = wrapper.result.current;
            const result1 = Array.isArray(value1) ? value1[1] : "BAD1";
            const result2 = Array.isArray(value2) ? value2[1] : "BAD2";

            // Assert
            expect(result1).not.toBe(result2);
        });

        it("should be a new function if the scope changes", () => {
            // Arrange
            const wrapper = renderHook(
                ({scope}: any) => useSharedCache("id", scope),
                {
                    initialProps: {scope: "scope"},
                },
            );

            // Act
            wrapper.rerender({scope: "new-scope"});
            const value1 = wrapper.result.all[wrapper.result.all.length - 2];
            const value2 = wrapper.result.current;
            const result1 = Array.isArray(value1) ? value1[1] : "BAD1";
            const result2 = Array.isArray(value2) ? value2[1] : "BAD2";

            // Assert
            expect(result1).not.toBe(result2);
        });

        it("should set the value in the cache", () => {
            // Arrange
            const wrapper = renderHook(() => useSharedCache("id", "scope"));
            const setValue = wrapper.result.current[1];

            // Act
            setValue("CACHED_VALUE");
            // Rerender so the hook retrieves this new value.
            wrapper.rerender();
            const result = wrapper.result.current[0];

            // Assert
            expect(result).toBe("CACHED_VALUE");
        });

        it.each`
            value
            ${undefined}
            ${null}
        `("should purge the value from the cache if $value", ({value}: any) => {
            // Arrange
            const wrapper = renderHook(() => useSharedCache("id", "scope"));
            const setValue = wrapper.result.current[1];
            setValue("CACHED_VALUE");

            // Act
            // Rerender so the result has the cached value.
            wrapper.rerender();
            setValue(value);
            // Rerender so the hook retrieves this new value.
            wrapper.rerender();
            const result = wrapper.result.current[0];

            // Assert
            expect(result).toBeNull();
        });
    });

    it("should share cache across all uses", () => {
        // Arrange
        const hook1 = renderHook(() => useSharedCache("id", "scope"));
        const hook2 = renderHook(() => useSharedCache("id", "scope"));
        hook1.result.current[1]("VALUE_1");

        // Act
        hook2.rerender();
        const result = hook2.result.current[0];

        // Assert
        expect(result).toBe("VALUE_1");
    });

    it.each`
        id
        ${"id1"}
        ${"id2"}
    `("should not share cache if scope is different", ({id}: any) => {
        // Arrange
        const hook1 = renderHook(() => useSharedCache("id1", "scope1"));
        const hook2 = renderHook(() => useSharedCache(id, "scope2"));
        hook1.result.current[1]("VALUE_1");

        // Act
        hook2.rerender();
        const result = hook2.result.current[0];

        // Assert
        expect(result).toBeNull();
    });

    it.each`
        scope
        ${"scope1"}
        ${"scope2"}
    `("should not share cache if id is different", ({scope}: any) => {
        // Arrange
        const hook1 = renderHook(() => useSharedCache("id1", "scope1"));
        const hook2 = renderHook(() => useSharedCache("id2", scope));
        hook1.result.current[1]("VALUE_1");

        // Act
        hook2.rerender();
        const result = hook2.result.current[0];

        // Assert
        expect(result).toBeNull();
    });
});
