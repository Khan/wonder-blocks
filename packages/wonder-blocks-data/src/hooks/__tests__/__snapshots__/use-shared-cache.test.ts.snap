// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#useSharedCache should throw if the id is  1`] = `[InvalidInputDataError: id must be a non-empty string]`;

exports[`#useSharedCache should throw if the id is [Function anonymous] 1`] = `[InvalidInputDataError: id must be a non-empty string]`;

exports[`#useSharedCache should throw if the id is 5 1`] = `[InvalidInputDataError: id must be a non-empty string]`;

exports[`#useSharedCache should throw if the id is null 1`] = `[InvalidInputDataError: id must be a non-empty string]`;

exports[`#useSharedCache should throw if the scope is  1`] = `[InvalidInputDataError: scope must be a non-empty string]`;

exports[`#useSharedCache should throw if the scope is [Function anonymous] 1`] = `[InvalidInputDataError: scope must be a non-empty string]`;

exports[`#useSharedCache should throw if the scope is 5 1`] = `[InvalidInputDataError: scope must be a non-empty string]`;

exports[`#useSharedCache should throw if the scope is null 1`] = `[InvalidInputDataError: scope must be a non-empty string]`;
